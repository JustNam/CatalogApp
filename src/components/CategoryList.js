import React, { Component } from 'react';
import { Tabs, Tab, TabPanel, TabList } from 'react-web-tabs';
import 'react-web-tabs/dist/react-web-tabs.css';
import ItemList from './ItemList';

class CategoryList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      categorySelected: false,
      categoryId: 0,
    };
  }

  componentDidMount() {
    this.props.getCategories();
  }

  getItems(e, categoryId) {
    e.preventDefault();
    this.setState({
      categorySelected: true,
    });
    // eslint-disable-next-line react/destructuring-assignment
    this.props.getItemsInCategory(categoryId);
    this.setState({ categoryId });
  }

  render() {
    // console
    const { category } = this.props;
    const { categoryId, categorySelected } = this.state;
    return (
      <div>
        <Tabs defaultTab="vertical-tab-one" vertical>
          <TabList>
            {category.data.map((category, index) => (
              <Tab
                tabFor={`vertical-tab-${index}`}
                key={category.id}
                onClick={e => this.getItems(e, category.id)}
              >
                <h3>{category.name}</h3>
              </Tab>
            ))}
            <br />
          </TabList>
          {!categorySelected && (
            <div className="welcome">
              Lets's explore our catalog!
              <br />
              Choose a category.
            </div>
          )}
          {category.data.map((category, index) => (
            <TabPanel tabId={`vertical-tab-${index}`} key={category.id}>
              {category.id === categoryId && <ItemList {...this.props} />}
            </TabPanel>
          ))}
        </Tabs>
      </div>
    );
  }
}

export default CategoryList;
