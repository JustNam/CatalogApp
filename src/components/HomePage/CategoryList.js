import React, { Component } from 'react';
import { Tabs, Tab, TabPanel, TabList } from 'react-web-tabs';
import 'react-web-tabs/dist/react-web-tabs.css';
import ItemList from 'components/HomePage/ItemList';
import NavigationBar from 'components/HomePage/NavigationBar';


class CategoryList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      categoryId: 0,
    };
  }

  async componentDidMount() {
    // Load the data of default category
    await this.props.getCategories().then(() => {
      const { category } = this.props;
      const { id } = category.data[0];
      this.props.getItemsInCategoryWithPagination(id, 1);
      this.setState({ categoryId: id });
    });
  }

  getItems(e, categoryId) {
    // Get items in chosen category
    e.preventDefault();
    this.props.getItemsInCategoryWithPagination(categoryId, 1);
    this.setState({ categoryId });
  }

  render() {
    // category variable contains the general information about categories
    const { category } = this.props;
    const { categoryId } = this.state;
    return (
      <div>
        <NavigationBar />
        <div className="container page">
          <Tabs defaultTab="vertical-tab-0" vertical>
            <TabList>
              {category.data.map((category, index) => (
                <Tab
                  tabFor={`vertical-tab-${index}`}
                  key={category.id}
                  onClick={e => this.getItems(e, category.id)}
                >
                  <h3>{category.name}</h3>
                </Tab>
              ))}
              <br />
            </TabList>
            {category.data.map((category, index) => (
              <TabPanel tabId={`vertical-tab-${index}`} key={category.id}>
                {category.id === categoryId && <ItemList {...this.props} />}
              </TabPanel>
            ))}
          </Tabs>
        </div>
      </div>
    );
  }
}

export default CategoryList;
