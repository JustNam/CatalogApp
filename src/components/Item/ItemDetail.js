import React, { Component } from 'react';
import { ButtonToolbar, Button } from 'react-bootstrap';
import EditItemModal from '../Modals/EditItemModal';
import SuccessModal from '../Modals/SuccessModal';
import ConfirmModal from '../Modals/ConfirmModal';
import NavigationBar from '../HomePage/NavigationBar';
import { historyWithRefresh } from '../../history';

class ItemDetail extends Component {
  constructor(props) {
    super(props);

    const { categoryId, itemId } = this.props.match.params;
    this.state = {
      showInfoModal: false,
      showSuccessModal: false,
      showConfirmModal: false,
      categoryId,
      itemId,
      backToHome: null,
      itemDetail: {},
    };
  }

  async componentDidMount() {
    // Prevent refresh from crashing
    const { categoryId, itemId } = this.props.match.params;
    await this.props.getItemInCategory(categoryId, itemId);
    const { item } = this.props;
    this.setState({ itemDetail: item.data[0] });
  }

  deleteItem = (e) => {
    e.preventDefault();
    const { categoryId, itemId } = this.state;
    this.props.deleteItemInCategory(categoryId, itemId).then(() => {
      this.handleConfirmClose();
      this.handleSuccessShow();
    });
  }

  handleInfoShow = () => {
    this.setState({ showInfoModal: true });
    this.setState({ backToHome: false });
  };

  handleInfoClose = () => {
    this.setState({ showInfoModal: false });
  };

  handleSuccessShow = () => {
    this.setState({ showSuccessModal: true });
  };

  handleSuccessClose = () => {
    const { categoryId, itemId } = this.state;
    this.setState({ showSuccessModal: false });
    if (this.state.backToHome) {
      historyWithRefresh.push('/categories');
    } else {
      historyWithRefresh.push(`/categories/${categoryId}/items/${itemId}`);
    }
  };

  handleConfirmShow = () => {
    this.setState({ showConfirmModal: true });
    this.setState({ backToHome: true });
  };

  handleConfirmClose = () => {
    this.setState({ showConfirmModal: false });
  };

  render() {
    const { itemDetail } = this.state;
    const userId = localStorage.getItem('userId');
    if (itemDetail.user) {
      const { showInfoModal, showSuccessModal, showConfirmModal, categoryId } = this.state;
      return (
        <div>
          <NavigationBar />
          <div className="container page">
            <div className="article-preview">
              <div className="article-meta">
                <div className="info">
                  <div className="author">
                    {itemDetail.user.username}
                  </div>
                  <span className="date">{`updated ${itemDetail.created_on.replace(',', ' at')}, created ${itemDetail.updated_on.replace(',', ' at')}`}</span>
                </div>
              </div>
              <div className="preview-link">
                <h1>{itemDetail.title}</h1>
                <p>{itemDetail.description}</p>
                {!itemDetail.description && <p><i>No description</i></p>}
                {(parseInt(userId) === parseInt(itemDetail.user.id))
                  && (
                  <span>
                    <ButtonToolbar className="modal-footer">
                      <Button
                        id="editButton"
                        variant="outline-success"
                        size="sm"
                        onClick={this.handleInfoShow}
                      >
                        Edit
                      </Button>
                      <Button
                        variant="outline-success"
                        size="sm"
                        onClick={this.handleConfirmShow}
                      >
                      Delete
                      </Button>
                    </ButtonToolbar>
                  </span>
                  )
                }
              </div>
            </div>
          </div>

          <EditItemModal
            currentCategoryId={categoryId}
            item={itemDetail}
            show={showInfoModal}
            handleClose={this.handleInfoClose}
            showSuccessModal={this.handleSuccessShow}
          />
          <SuccessModal show={showSuccessModal} handleClose={this.handleSuccessClose} />
          <ConfirmModal
            show={showConfirmModal}
            handleConfirm={this.deleteItem}
            handleClose={this.handleConfirmClose}
          />
        </div>
      );
    }

    return <div><NavigationBar /></div>;
  }
}
export default ItemDetail;