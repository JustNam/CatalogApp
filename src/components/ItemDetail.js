import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

class ItemDetail extends Component {
  componentDidMount() {
    // eslint-disable-next-line react/destructuring-assignment
    const { categoryId, itemId } = this.props.match.params;
    console.log(categoryId);
    console.log(itemId);
    // eslint-disable-next-line react/destructuring-assignment
    this.props.getItemInCategory(categoryId, itemId);
  }

  render() {
    const { item } = this.props;
    const itemDetail = item.data[0];
    if (itemDetail) {
      return (
        <div>
          <div className="article-preview">
            <div className="article-meta">
              <div className="info">
                <div className="author">
                  {itemDetail.user.username}
                </div>
                <span className="date">{`updated ${itemDetail.created_on.replace(',', ' at')}, created ${itemDetail.updated_on.replace(',', ' at')}`}</span>
              </div>
            </div>
            <div className="preview-link" href="#article/books-4kbydz">
              <h1>{itemDetail.title}</h1>
              <p>{itemDetail.description}</p>
              {!itemDetail.description && <p><i>No description</i></p>}
            </div>
          </div>
        </div>
      );
    }
    return <div />;
  }
}

const mapStateToProps = state => ({});
const mapDispatchToProps = dispatch => ({});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ItemDetail));
