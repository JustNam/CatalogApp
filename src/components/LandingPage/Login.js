import React, { Component } from 'react';
import { Redirect, Link } from 'react-router-dom';
import { Button } from 'react-bootstrap';

class LoginPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      username: '',
      password: '',
      redirect: false,
      error: false,
    };
  }

  handleUsernameChange = (e) => {
    this.setState({ username: e.target.value });
  };

  handlePasswordChange = (e) => {
    this.setState({ password: e.target.value });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { username, password } = this.state;
    this.props.login(username, password).then((response) => {
      if (response.payload.error) {
        this.setState({ error: true });
      } else {
        this.setState({ redirect: true });
      }
    });
  };

  render() {
    const { username, password, redirect, error } = this.state;
    if (redirect) {
      return <Redirect to="/categories" />;
    }
    return (
      <div className="container">
        <div className="row">
          <div className="col-sm-9 col-md-7 col-lg-5 mx-auto">
            <div className="card card-signin my-5">
              <div className="card-body">
                <h5 className="card-username text-center">Catalog App</h5>
                <form className="form-signin">
                  {this.props.location.signUpSuccess && (
                    <div className="success-message">
                      Sign up successfully, log in with your new account!
                    </div>
                  )}
                  {error && (
                    <div className="invalid-credential">
                      Invalid username or password
                    </div>
                  )}
                  <div className="form-label-group">
                    <input
                      type="text"
                      id="inputUsername"
                      className="form-control"
                      placeholder="Email address"
                      value={username}
                      onChange={this.handleUsernameChange}
                      required
                    />
                    <label htmlFor="inputUsername">Username</label>
                  </div>

                  <div className="form-label-group">
                    <input
                      type="password"
                      id="inputPassword"
                      className="form-control"
                      placeholder="Password"
                      value={password}
                      onChange={this.handlePasswordChange}
                      required
                    />
                    <label htmlFor="inputPassword">Password</label>
                  </div>

                  <button
                    className="btn btn-lg btn-primary btn-block text-uppercase"
                    onClick={this.handleSubmit}
                    type="submit"
                  >
                    Log in
                  </button>
                  <Button
                    href="/signup"
                    className="btn btn-lg btn-outline-secondary btn-block text-uppercase"
                    type="button"
                  >
                    Sign up
                  </Button>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
export default LoginPage;
