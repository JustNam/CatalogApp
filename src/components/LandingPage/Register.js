import React, { Component } from 'react';
import { Form, Button } from 'react-bootstrap';
import { Redirect } from 'react-router-dom';
import { validateUsername } from '../../utilities/validate';

class RegisterPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isValidRequest: '',
      username: {
        value: '',
        errorMessages: [],
        isValid: false,
      },
      password: '',
      confirmPassword: '',
      passwordMatch: false,
      usernameInputProps: {
        isValid: null,
        isInvalid: null,
      },
      passwordInputProps: {
        isValid: null,
        isInvalid: null,
      },
      redirect: false,
    };
  }

  updateUsernameInputProps = (validity) => {
    this.setState({
      usernameInputProps: {
        isValid: validity,
        isInvalid: !validity,
      },
    });
  };

  resetUsernameInputProps = () => {
    this.setState({
      usernameInputProps: {
        isInvalid: null,
        isValid: null,
      },
    });
  };

  updatePasswordInputProps = (validity) => {
    this.setState({
      passwordInputProps: {
        isValid: validity,
        isInvalid: !validity,
      },
    });
  };

  resetPasswordInputProps = () => {
    this.setState({
      passwordInputProps: {
        isValid: null,
        isInvalid: null,
      },
    });
  };

  validateUsernameData = (username) => {
    const usernameValidation = validateUsername(username);
    const { passwordMatch } = this.state;
    this.setState({
      username: {
        ...usernameValidation,
      },
      isValidRequest: usernameValidation.isValid && passwordMatch,
    });
    this.updateUsernameInputProps(usernameValidation.isValid);
  };

  // Validate username and trigger update of username object
  handleUsernameChange = (e) => {
    const value = e.target.value;
    if (value === '') {
      // Reset input props, clear state when the input is empty
      this.setState({
        username: {
          value: '',
          errorMessages: [],
          isValid: false,
        },
      });
      this.resetUsernameInputProps();
    } else {
      const newUsername = {
        ...this.state.username,
        value,
      };
      this.validateUsernameData(newUsername);
    }
  };

  // Update password value, compare with confirmPassword and update validadtions
  handlePasswordChange = (e) => {
    const { confirmPassword, username } = this.state;
    const password = e.target.value;
    if (password === '') {
      this.resetPasswordInputProps();
      this.setState({ password: '' });
    } else {
      const passwordMatch = password === confirmPassword;
      this.setState({
        password,
        passwordMatch,
        isValidRequest: username.isValid && passwordMatch,
      });
      if (confirmPassword) {
        this.updatePasswordInputProps(passwordMatch);
      }
    }
  };

  // Update confirmPassword value, compare with password and update validadtions
  handleConfirmPasswordChange = (e) => {
    const { password, username } = this.state;
    const confirmPassword = e.target.value;
    if (confirmPassword === '') {
      this.resetPasswordInputProps();
      this.setState({ confirmPassword: '' });
    } else {
      const passwordMatch = (password === confirmPassword);
      this.setState({
        confirmPassword,
        passwordMatch,
        isValidRequest: username.isValid && passwordMatch,
      });
      if (password) {
        this.updatePasswordInputProps(passwordMatch);
      }
    }
  };

  handleSubmit = async (e) => {
    e.preventDefault();
    const { username, password, isValidRequest } = this.state;
    if (isValidRequest) {
      // Check if payload contains "error" key and udpate errors in state
      await this.props.signUp(username.value, password)
        .then((response) => {
          if (response.payload.error) {
            this.setState({
              username: {
                ...username,
                errorMessages: ['The username already exists'],
              },
            });
            this.updateUsernameInputProps(false);
            this.setState({ test: 'true' });
          } else {
            this.setState({ test: 'false' });
            this.setState({ redirect: true });
          }
        });
    }
  };

  render() {
    const {
      username,
      password,
      confirmPassword,
      redirect,
      usernameInputProps,
      passwordInputProps,
      test,
    } = this.state;
    if (redirect) {
      return (
        <Redirect to={{
          pathname: '/login',
          signUpSuccess: true,
        }}
        />
      );
    }
    return (
      <div className="container">
        <div>{test}</div>
        <div className="row">
          <div className="col-sm-9 col-md-7 col-lg-5 mx-auto">
            <div className="card card-signin my-5">
              <div className="card-body">
                <h5 className="card-username text-center">Catalog App</h5>
                <div className="minor-title">Register new account</div>
                <form
                  className="form-signin"
                  autoComplete="off"
                  onSubmit={this.handleSubmit}
                >
                  <div className="form-label-group">
                    <Form.Control
                      type="text"
                      id="inputUsername"
                      placeholder="Username"
                      minLength="6"
                      maxLength="30"
                      value={username.value}
                      onChange={this.handleUsernameChange}
                      {...usernameInputProps}
                      required
                    />
                    <label htmlFor="inputUsername">Username</label>
                    <Form.Control.Feedback type="invalid">
                      {username.errorMessages.map((errorMessage, index) => (
                        <div key={index}>{errorMessage}</div>
                      ))}
                    </Form.Control.Feedback>

                  </div>
                  <div className="form-label-group">

                    <Form.Control
                      type="password"
                      id="inputPassword"
                      className="form-control"
                      placeholder="Password"
                      value={password}
                      onChange={this.handlePasswordChange}
                      minLength="6"
                      {...passwordInputProps}
                      required
                    />
                    <label htmlFor="inputPassword">Password</label>
                  </div>

                  <div className="form-label-group">
                    <Form.Control
                      type="password"
                      id="inputConfirmPassword"
                      className="form-control"
                      placeholder="Confirm Password"
                      value={confirmPassword}
                      onChange={this.handleConfirmPasswordChange}
                      required
                      {...passwordInputProps}
                    />
                    <label htmlFor="inputConfirmPassword">
                      Confirm Password
                    </label>
                    <Form.Control.Feedback type="invalid">
                    Confirm password does not match
                    </Form.Control.Feedback>
                  </div>

                  <button
                    className="btn btn-lg btn-primary btn-block text-uppercase"
                    type="submit"
                  >
                    Submit
                  </button>
                  <Button
                    href="/login"
                    className="btn btn-lg btn-outline-secondary btn-block text-uppercase"
                    type="button"
                  >
                    Back to login
                  </Button>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default RegisterPage;
