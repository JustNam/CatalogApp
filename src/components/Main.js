import React, { Component } from 'react';
import { Route, Switch, Redirect } from 'react-router-dom';
import CategoryList from './HomePage/CategoryList';
import ItemDetail from './Item/ItemDetail';
import LoginPage from './LandingPage/Login';
import RegisterPage from './LandingPage/Register';

class Main extends Component {
  render() {
    // Redirect to /categories if the URL does not match any route
    const userRoutes = [
      <Route
        exact
        key="item-detail"
        path="/categories/:categoryId/items/:itemId"
        render={params => <ItemDetail {...this.props} {...params} />}
      />,
      <Route
        exact
        key="category-list"
        path="/categories"
        render={() => <CategoryList {...this.props} />}
      />,
      <Redirect key="redirect-category" to="/categories" />,
    ];
    // Redirect to /login if the URL does not match with any route and current actor is guest
    const guestRoutes = [
      <Route
        path="/signup"
        key="register"
        render={() => (
          <RegisterPage {...this.props} />
        )}
      />,
      <Route
        path="/login"
        key="login"
        render={() => (
          <LoginPage {...this.props} />
        )}
      />,
      <Redirect key="redirect-login" to="/login" />,
    ];
    if (localStorage.getItem('accessToken')) {
      return (
        <Switch>
          {userRoutes}
        </Switch>
      );
    }
    return (
      <Switch>
        {guestRoutes}
      </Switch>
    );
  }
}

export default Main;
