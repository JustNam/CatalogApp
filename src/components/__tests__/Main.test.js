import React from 'react';
import { Route } from 'react-router-dom';
import { shallow } from 'enzyme';
import Main from '../Main';
import Register from '../LandingPage/Register';
import Login from '../LandingPage/Login';
import ItemDetail from '../Item/ItemDetail';
import CategoryList from '../HomePage/CategoryList';

let pathMap = {};
describe('Main for guests', () => {
  beforeAll(() => {
    const component = shallow(<Main />);
    pathMap = component.find(Route).reduce((accumulator, route) => {
      const routeProps = route.props();
      accumulator[routeProps.path] = routeProps.render().type;
      return accumulator;
    }, {});
  });
  let wrapper;
  it('It should render correctly', () => {
    wrapper = shallow(<Main />);
    expect(wrapper).toMatchSnapshot();
  });
  it('It should show components for guests', () => {
    expect(pathMap['/signup']).toBe(Register);
    expect(pathMap['/login']).toBe(Login);
  });
});
describe('Main for users', () => {
  const pathMap = {};
  beforeAll(() => {
    localStorage.setItem('accessToken', 'test');
    const component = shallow(<Main />);
    const routes = component.find(Route);
    routes.forEach((route, index) => {
      if (index !== routes.length) {
        const routeProps = route.props();
        pathMap[routeProps.path] = routeProps.render().type;
      }
    });
  });
  let wrapper;
  it('It should render correctly', () => {
    wrapper = shallow(<Main />);
    expect(wrapper).toMatchSnapshot();
  });
  it('It should show components for guests', () => {
    expect(pathMap['/categories/:categoryId/items/:itemId']).toBe(ItemDetail);
    expect(pathMap['/categories']).toBe(CategoryList);
  });
});