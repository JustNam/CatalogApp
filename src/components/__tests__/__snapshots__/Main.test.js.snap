// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main for guests It should render correctly 1`] = `ShallowWrapper {}`;

exports[`Main for users It should render correctly 1`] = `ShallowWrapper {}`;
